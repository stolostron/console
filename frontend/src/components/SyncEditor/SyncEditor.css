:root {
    --sy-c-code-editor__code--PaddingTop: 0;
    --sy-c-code-editor__code--PaddingRight: 0;
    --sy-c-code-editor__code--PaddingBottom: 0;
    --sy-c-code-editor__code--PaddingLeft: 0;
}

.sy-c-code-editor__title {
    display: flex;
    align-items: center;
    padding: 0 12px;
}

.sync-editor__container {
    width: 100%;
    height: 100%;
    overflow: hidden;
}

.sync-editor__container .pf-c-code-editor__header {
    font-size: 14px;
    font-weight: 600;
    color: #ededed;
    background-color: #25282c;
    border-bottom: 1px solid #55585c;
}

.sync-editor__container .pf-c-code-editor__controls {
    width: 100%;
    justify-content: space-between;
}

.sync-editor__container .pf-c-code-editor__header::before {
    border: none;
}

.sync-editor__container .pf-c-code-editor__controls .pf-c-button.pf-m-control {
    background-color: #25282c;
}

.sync-editor__container .pf-c-button.pf-m-plain,
.pf-c-code-editor__controls .pf-c-button.pf-m-control:not([disabled]) {
    color: #ededed;
}

.sync-editor__container .pf-c-button::after {
    border: none;
    border-left: 1px solid #55585c;
}

.sync-editor__container .pf-c-button:hover {
    color: #ededed;
    background-color: #25282c;
}

.sync-editor__container .pf-c-code-editor__main {
    border: none;
}

.sync-editor__container .pf-c-code-editor__code {
    padding: var(--sy-c-code-editor__code--PaddingTop) var(--sy-c-code-editor__code--PaddingRight)
        var(--sy-c-code-editor__code--PaddingBottom) var(--sy-c-code-editor__code--PaddingLeft);
}

.protectedDecoration {
    color: #b0b0b0 !important;
    user-select: auto;
}

.errorDecoration {
    margin-left: 6px;
    height: 14px !important;
    width: 14px !important;
    cursor: pointer;
    z-index: 1000;
    background: url("data:image/svg+xml;utf8, <svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16' width='14' height='14'><g fill='red'><circle cx='8' cy='8' r='8' fill='white' fill-opacity='1'/><path d='M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zM5.1 13.3L3.5 12 11 2.6l1.5 1.2-7.4 9.5z'/> </g></svg>");
}
.insertedLineDecoration {
    background: #87d4ecab;
    width: 6px !important;
    margin-left: 12px;
}

.customLineDecoration {
    background: blue;
    width: 6px !important;
    margin-left: 12px;
}
