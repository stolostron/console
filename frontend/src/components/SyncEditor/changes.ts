/* Copyright Contributors to the Open Cluster Management project */
import YAML from 'yaml'
import { diff } from 'deep-diff'
import { get, isEmpty, keyBy, isEqual } from 'lodash'
import { getPathArray } from './synchronize'
import { normalize } from './reconcile'
import { MappingType } from './process'

export interface ChangeType {
    $t: string // type of change (N, E)
    $u?: any // the value the user has changed it to
    $y?: boolean // an element was added to the array at $p
    $f?: any // the previous value in the form when the user has edited
    $a: string | string[] // the path to the value in a mapped object
    $p: string | string[] // the path to the value in a parsed object
}

export interface FormattedChangeType {
    type: string
    line: number
    path: string | string[]
    previous?: string
    latest?: string | string[]
    length: number
    reveal?: () => void
}

export const getFormChanges = (
    errors: any[],
    change: {
        parsed: { [name: string]: any[] }
        mappings: { [name: string]: any[] }
    },
    userEdits: any[],
    comparison: { [name: string]: any[] },
    lastChange?: {
        parsed: { [name: string]: any[] }
        mappings: { [name: string]: any[] }
    },
    lastComparison?: { [name: string]: any[] }
) => {
    let yamlChanges = []
    let remainingEdits = []

    // changes to yaml
    if (errors.length === 0 || !errors.every(({ isWarning }) => !isWarning)) {
        if (lastChange && lastComparison) {
            yamlChanges = getChanges(false, change, lastChange, comparison, lastComparison)

            // remove any form changes on top of user changes (for decoration purposes--reconcile prevents the yaml change)
            // remove any user changes which are now the same as the form
            if (!isEmpty(userEdits)) {
                // changes made by form
                const changeMap = keyBy(yamlChanges, (edit) => {
                    return JSON.stringify(edit.$y ? edit.$p.slice(0, -1) : edit.$p)
                })

                // changes made by user
                remainingEdits = userEdits.filter((edit: ChangeType) => {
                    const { $u, $y, $p } = edit
                    const val = get(change.parsed, $p) as MappingType
                    // if there's no value at this path anymore, just filter out this user edit (may have deleted)
                    if (val !== undefined) {
                        // use any form change on top of a user edit
                        const key = JSON.stringify($p)
                        const chng = changeMap[key]
                        if (chng) {
                            // user edit and form change conflict at this path, so delete form change
                            delete changeMap[key]
                            // if change equals the user edit, also delete the user edit
                            if ((Array.isArray(val) && $y ? val.includes($u) : isEqual($u, val)) || (!$u && !val)) {
                                return false
                            }
                        }
                        return true
                    }
                    return false
                })
                yamlChanges = Object.values(changeMap)
            }
        }
    }
    return { yamlChanges, remainingEdits }
}

export const getUserChanges = (
    errors: any[],
    change: {
        parsed: { [name: string]: any[] }
        mappings: { [name: string]: any[] }
    },
    lastUserEdits: any[],
    comparison: { [name: string]: any[] },
    lastChange?: {
        parsed: { [name: string]: any[] }
        mappings: { [name: string]: any[] }
    },
    lastComparison?: { [name: string]: any[] }
) => {
    let changes = [] //lastUserEdits

    // changes to yaml
    if (errors.length === 0 || !errors.every(({ isWarning }) => !isWarning)) {
        if (lastChange && lastComparison) {
            // comparison is always made between current user changes and last yaml created by form
            changes = getChanges(true, change, lastChange, comparison, lastComparison)

            // merge these newest user edits with older user edits
            // lastUserEdits wil only exist if user typed in editor, then form created a change, and now user is again typing in editor
            // newest changes just reflect changes between what user types and what's currently in yaml
            // but after a form change, any previous user edits are now reconciled into the yaml
            // so so they won't show up in the newest changes---the only place we remember old edits are in lastUserEdits
            if (!isEmpty(lastUserEdits)) {
                // create a map of newest user changes
                const changeMap = keyBy(changes, (edit) => {
                    return JSON.stringify(edit.$p)
                })
                // for each older user edit...
                lastUserEdits.forEach((oldEdit) => {
                    const { $a, $f, $p } = oldEdit
                    // a form change has canceled out this user edit
                    // (the form has changed the value to equal what the user edit )
                    const val = get(change.mappings, $a) as unknown as MappingType
                    // if there's no value at this path anymore, ignore this old edit
                    // ---the form has probably deleted where the edit was from the yaml
                    if (val) {
                        const key = JSON.stringify($p)
                        const chng = changeMap[key]
                        if (chng) {
                            // user is editing the same path as an old edit
                            // if the change they make is the same value as that old edit
                            // this change and the old edit are unnecessary
                            if (val.$v === $f) {
                                delete changeMap[key]
                            } else {
                                // else if user is editing the same spot, the newest change's $f won't be the right value
                                // because it's comparing against the latest yaml which has incorporated the previous user edits
                                // the only place that remembers the original $f value is the old edit
                                chng.$f = $f
                            }
                        } else {
                            // else user isn't typing the same spot as this old edit
                            // and the newest changes don't know about the old edit
                            // so add it to newest changes
                            // however if the form made a change that now equals this old edit value
                            // ignore it
                            if (val.$v !== $f) {
                                changeMap[key] = oldEdit
                            }
                        }
                    }
                })
                // changeMap now has newest changes, tweaked new changes, and old edits
                changes = Object.values(changeMap)
            }
        }
    }
    return changes
}

const getChanges = (
    isCustomEdit: boolean,
    change: {
        parsed: { [name: string]: any[] }
        mappings: { [name: string]: any[] }
    },
    lastChange: {
        parsed: { [name: string]: any[] }
        mappings: { [name: string]: any[] }
    },
    comparison: { [name: string]: any[] },
    lastComparison: { [name: string]: any[] }
) => {
    const changes: any[] = []
    const ignorePaths: any = []
    normalize(lastComparison, comparison)
    const diffs = diff(lastComparison, comparison)
    if (diffs) {
        diffs.forEach((diff: any) => {
            const { path, item, lhs, rhs } = diff
            let { kind } = diff
            if (path && path.length) {
                let pathArr = getPathArray(path)
                const synced = (kind === 'D' || kind === 'E') && lhs && !rhs ? lastChange.mappings : change.mappings
                let obj: any = get(synced, pathArr)
                if (obj) {
                    if (obj.$v || obj.$v === false) {
                        // convert A's and E's into 'N's
                        switch (kind) {
                            case 'E': {
                                if (obj.$l > 1 && !lhs && rhs) {
                                    // convert edit to new if multiple lines were added
                                    kind = 'N'
                                    obj = { $r: obj.$r + 1, $l: obj.$l - 1 }
                                }
                                break
                            }
                            case 'A': {
                                switch (item.kind) {
                                    case 'D':
                                        // if array delete, ignore any other edits within array
                                        // edits are just the comparison of other array items
                                        ignorePaths.push(path.join('/'))
                                        break
                                }
                                break
                            }
                        }
                    } else if (obj.$l > 1 && path.length > 0 && kind !== 'D') {
                        kind = 'N'
                        path.pop()
                        pathArr = getPathArray(path)
                        obj = get(change.mappings, pathArr)
                    }

                    // if array delete, ignore any other edits within array
                    // edits are just the comparison of other array items
                    if (ignorePaths.length > 0) {
                        const tp = path.join('/')
                        if (
                            ignorePaths.some((p: any) => {
                                return tp.startsWith(p)
                            })
                        ) {
                            // ignore any edits within an array that had an item deleted
                            kind = 'D'
                        }
                    }

                    let chng: ChangeType
                    switch (kind) {
                        case 'A': {
                            if (item.kind === 'N') {
                                chng = { $t: 'N', $a: pathArr, $p: path, $y: true }
                                if (isCustomEdit) {
                                    chng.$u = item.rhs
                                    chng.$f = 'new'
                                } else {
                                    if (Array.isArray(obj.$v)) {
                                        obj.$v.some((itm: { $v: any; rhs: any; $k: any }) => {
                                            if ((!itm.$v && item.rhs === '') || itm.$v === item.rhs) {
                                                chng.$a = [...pathArr, '$v', itm.$k]
                                                chng.$p = [...path, itm.$k]
                                                return true
                                            }
                                            return false
                                        })
                                    }
                                }
                                changes.push(chng)
                            }
                            break
                        }
                        case 'E': {
                            // edited
                            if ((obj.$v || obj.$v === false) && rhs !== undefined) {
                                chng = { $t: 'E', $a: pathArr, $p: path }
                                if (isCustomEdit) {
                                    chng.$u = rhs // what user changed it to
                                    chng.$f = lhs // what form had it as
                                }
                                changes.push(chng)
                            }
                            break
                        }
                        case 'N': // new
                            chng = { $t: 'N', $a: pathArr, $p: path }
                            if (isCustomEdit) {
                                chng.$u = rhs || obj.$v // what user changed it to
                                chng.$f = 'new'
                            }
                            changes.push(chng)
                            break
                    }
                }
            }
        })
    }
    return changes
}

export const formatChanges = (
    editor: { revealLineInCenter: (arg0: any) => void; setSelection: (arg0: any) => void },
    monaco: { Selection: new (arg0: any, arg1: number, arg2: any, arg3: number) => any },
    changes: any[],
    changeWithoutSecrets: { mappings: any; parsed: any; resources?: any[] },
    syncs: unknown
) => {
    const syncPathSet = new Set()
    if (Array.isArray(syncs)) {
        syncs.forEach(({ path }) => {
            syncPathSet.add(getPathArray(path).join('/'))
        })
    }
    changes = changes
        .filter((change: ChangeType) => {
            // don't include change if it's already being sent to form
            // or if the value is undefined
            return (
                !syncPathSet.has((change.$a as string[]).join('/')) &&
                get(changeWithoutSecrets.parsed, change.$p) !== undefined
            )
        })
        .map((change: ChangeType) => {
            const obj = get(changeWithoutSecrets.parsed, change.$p)
            const objVs = get(changeWithoutSecrets.mappings, change.$a)
            const formatted: FormattedChangeType = {
                type: change.$t,
                line: objVs?.$r ?? 1,
                path: change.$p,
                length: objVs.$l,
            }
            switch (change.$t) {
                case 'N':
                    formatted.latest = YAML.stringify({ [objVs.$k]: obj }, { indent: 4 })
                        .trim()
                        .split('\n')
                    break
                case 'E':
                    formatted.latest = objVs?.$v ?? ''
                    formatted.previous = objVs.$s ? formatted.latest : change.$f
                    break
            }
            return formatted
        })
        .sort((a: { line: number }, b: { line: number }) => {
            return a.line - b.line
        })
    changes = consolidate(changes)
    changes.forEach((change) => {
        change.reveal = () => {
            editor.revealLineInCenter(change.line)
            editor.setSelection(new monaco.Selection(change.line, 1, change.line + change.length, 1))
        }
    })
    return changes
}

const consolidate = (changes: FormattedChangeType[]) => {
    let lastChange: FormattedChangeType
    return changes.filter((change: FormattedChangeType) => {
        if (lastChange) {
            const { type, line, length } = lastChange
            if (type === 'N') {
                switch (change.type) {
                    case 'N':
                        // if last change was new and this change is new
                        // and last ends where this change begins
                        // consolidate this change into the last change
                        if (change.latest && lastChange.latest) {
                            if (line + length === change.line) {
                                if (!Array.isArray(change.latest)) {
                                    change.latest = [change.latest]
                                }
                                if (!Array.isArray(lastChange.latest)) {
                                    lastChange.latest = [lastChange.latest]
                                }
                                lastChange.latest = [...lastChange.latest, ...change.latest]
                                lastChange.length += change.length
                                return false
                            }
                        }
                        lastChange = change
                        break
                    case 'E':
                        // if last change was new and this change is edit
                        // and edit change falls with new
                        // just ignore this change
                        if (change.line < line + length) {
                            return false
                        }
                        break
                }
            } else {
                lastChange = change
            }
        } else {
            lastChange = change
        }
        return true
    })
}
