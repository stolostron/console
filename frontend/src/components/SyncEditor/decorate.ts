/* Copyright Contributors to the Open Cluster Management project */
import { get, capitalize } from 'lodash'

export const decorate = (
    isCustomEdit: boolean,
    editorHasFocus: boolean,
    editorRef: any,
    monacoRef: any,
    errors: any[],
    changes: any[],
    change: {
        parsed: { [name: string]: any[] }
        mappings: { [name: string]: any[] }
    },
    userEdits: any[],
    protectedRanges: any[],
    filteredRows: number[]
) => {
    const decorations: any[] = []
    const squigglyTooltips: any[] = []

    // errors/warnings
    addErrorDecorations(monacoRef, errors, decorations, squigglyTooltips)

    // add change decorations
    addChangeDecorations(isCustomEdit, monacoRef, changes, change, decorations)

    // if form is making changes, layer any editor changes decorations on top of form changes
    if (userEdits.length) {
        addChangeDecorations(true, monacoRef, userEdits, change, decorations)
    }

    // add protected decorations
    addProtectedDecorations(monacoRef, protectedRanges, decorations)

    // add filter row toggle decorations
    addFilteredDecorations(monacoRef, filteredRows, decorations)

    // add decorations to editor
    const handles = getResourceEditorDecorations(editorRef).map((decoration: { id: any }) => decoration.id)
    editorRef.current.deltaDecorations(handles, decorations)

    // scroll to best line to show
    if (!editorHasFocus) {
        scrollToChangeDecoration(editorRef, errors, decorations)
    }

    return squigglyTooltips
}

const addProtectedDecorations = (monacoRef: any, protectedRanges: any[], decorations: any[]) => {
    protectedRanges?.forEach((range) => {
        const start = range.startLineNumber
        const end = range.endLineNumber - 1
        decorations.push({
            range: new monacoRef.current.Range(start, 1, end, 132),
            options: {
                inlineClassName: 'protectedDecoration',
                description: 'resource-editor',
            },
        })
    })
}

const addFilteredDecorations = (monacoRef: any, filteredRows: any[], decorations: any[]) => {
    filteredRows?.forEach((row) => {
        decorations.push({
            range: new monacoRef.current.Range(row, 0, row, 132),
            options: {
                after: { content: '\u200b', inlineClassName: 'inline-folded' },
                description: 'resource-editor',
            },
        })
    })
}

const addErrorDecorations = (monacoRef: any, errors: any[], decorations: any[], squigglyTooltips: any[]) => {
    errors.forEach((error: { linePos: any; message: any; isWarning: boolean }) => {
        const { linePos, message, isWarning } = error
        const start = linePos?.start?.line ?? 0
        if (linePos && start !== 0) {
            // error in margin
            decorations.push({
                range: new monacoRef.current.Range(start, 0, start, 132),
                options: {
                    isWholeLine: true,
                    glyphMarginClassName: isWarning ? 'warningDecoration' : 'errorDecoration',
                    glyphMarginHoverMessage: { value: '```html\n' + capitalize(message) + ' \n```' },
                    overviewRuler: { color: isWarning ? '#ffff00' : '#ff0000', position: 4 },
                    minimap: { color: isWarning ? '#ffff0060' : '#ff000060', position: 1 },
                    description: 'resource-editor',
                },
            })

            // squiggly line under error
            const range = new monacoRef.current.Range(
                start,
                linePos?.start?.col ?? 0,
                linePos?.end?.line ?? start,
                linePos?.end?.col ?? 132
            )
            decorations.push({
                range,
                options: {
                    className: isWarning ? 'squiggly-warning' : 'squiggly-error',
                },
            })
            squigglyTooltips.push({
                range,
                message: capitalize(message.replace(/\^*/g, '').replace(/\n/g, '  ')),
            })
        }
        errors.push({ linePos, message })
    })
}

const addChangeDecorations = (
    isCustomEdit: boolean,
    monacoRef: any,
    changes: any[],
    change: {
        parsed: { [name: string]: any[] }
        mappings: { [name: string]: any[] }
    },
    decorations: any[]
) => {
    changes.forEach((chng) => {
        const { $t, $a, $f } = chng
        const obj: any = get(change.mappings, $a)
        if (obj) {
            decorations.push({
                range: new monacoRef.current.Range(obj.$r, 0, obj.$r + ($t === 'N' ? obj.$l - 1 : 0), 0),
                options: {
                    isWholeLine: true,
                    linesDecorationsClassName: isCustomEdit ? 'customLineDecoration' : 'insertedLineDecoration',
                    overviewRuler: isCustomEdit ? { color: '#0000ff', position: 1 } : {},
                    minimap: { color: isCustomEdit ? '#0000ff' : '#c0c0ff', position: 2 },
                    description: 'resource-editor',
                },
            })
            if ($f != null && $f.toString().length < 32 && !obj.$s) {
                decorations.push({
                    range: new monacoRef.current.Range(obj.$r, 0, obj.$r, 132),
                    options: {
                        after: { content: `  # ${$f}`, inlineClassName: 'protectedDecoration' },
                        description: 'resource-editor',
                    },
                })
            }
        }
    })
}

export const getResourceEditorDecorations = (editorRef: any) => {
    // clear resource-editor decorations
    const model = editorRef.current?.getModel()
    return model
        .getAllDecorations()
        .filter(
            (decoration: { options: { className: string; description: string } }) =>
                decoration?.options?.className?.startsWith('squiggly-') ||
                decoration?.options?.description === 'resource-editor'
        )
}

const scrollToChangeDecoration = (editorRef: any, errors: any[], decorations: any[]) => {
    const editor = editorRef.current
    const visibleRange = editor.getVisibleRanges()[0]
    if (visibleRange) {
        // if any errors and not in visible range, and first error isn't visible, scroll to it
        const errorLine = errors.length !== 0 ? errors[0]?.linePos?.start?.linePos : undefined
        if (errorLine && !visibleRange.containsPosition({ lineNumber: errorLine, column: 1 })) {
            setTimeout(() => {
                editor.revealLineInCenter(errorLine)
            })
        } else if (decorations.length) {
            // if visible range doesn't show any change decorations, scroll to first change decoration
            const changeDecorations = decorations.filter(
                (decoration) => decoration.options.linesDecorationsClassName === 'insertedLineDecoration'
            )
            if (
                changeDecorations.length &&
                !changeDecorations.some((decoration) => {
                    return visibleRange.containsPosition(decoration?.range.getStartPosition())
                })
            ) {
                setTimeout(() => {
                    editor.revealLineInCenter(changeDecorations[0]?.range.getStartPosition()?.lineNumber)
                })
            }
        }
    }
}
