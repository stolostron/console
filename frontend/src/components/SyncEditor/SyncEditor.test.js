/* Copyright Contributors to the Open Cluster Management project */
'use strict'

import React from 'react'
import { SyncEditor } from './SyncEditor'
import { render } from '@testing-library/react'
import schema from '../../routes/Governance/policies/schema.json'

describe('SyncEditor component', () => {
  afterAll(() => {
    jest.resetAllMocks()
  })

  it('renders as expected', () => {
    const update = jest.fn()
    const Component = (props) => {
      return <SyncEditor {...props} />
    }
    const { asFragment } = render(
      <Component
        editorTitle={'Policy YAML'}
        variant="toolbar"
        resources={[
          {
            apiVersion: 'policy.open-cluster-management.io/v1',
            kind: 'Policy',
            metadata: {
              name: 'policy-pod',
              namespace: 'default',
            },
            spec: {
              dependencies: [
                {
                  apiVersion: 'policy.open-cluster-management.io/v1',
                  compliance: 'Compliant',
                  kind: 'Policy',
                  name: 'case-test-policy',
                  namespace: 'default',
                },
              ],
              disabled: false,
              'policy-templates': [
                {
                  objectDefinition: {
                    apiVersion: 'policy.open-cluster-management.io/v1',
                    kind: 'ConfigurationPolicy',
                    metadata: {
                      name: 'policy-pod-1',
                    },
                    spec: {},
                  },
                },
              ],
            },
          },
        ]}
        schema={schema}
        onEditorChange={(changes) => {
          update(changes?.resources)
        }}
        mock={true}
        readonly={true}
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })
  it('renders as expected when a dependency compliance is invalid', () => {
    const update = jest.fn()
    const Component = (props) => {
      return <SyncEditor {...props} />
    }
    const { asFragment } = render(
      <Component
        editorTitle={'Policy YAML'}
        variant="toolbar"
        resources={[
          {
            apiVersion: 'policy.open-cluster-management.io/v1',
            kind: 'Policy',
            metadata: {
              name: 'policy-pod',
              namespace: 'default',
            },
            spec: {
              dependencies: [
                {
                  apiVersion: 'policy.open-cluster-management.io/v1',
                  compliance: 'notAComplianceType',
                  kind: 'Policy',
                  name: 'case-test-policy',
                  namespace: 'default',
                },
              ],
              disabled: false,
              'policy-templates': [
                {
                  objectDefinition: {
                    apiVersion: 'policy.open-cluster-management.io/v1',
                    kind: 'ConfigurationPolicy',
                    metadata: {
                      name: 'policy-pod-1',
                    },
                    spec: {},
                  },
                },
              ],
            },
          },
        ]}
        schema={schema}
        onEditorChange={(changes) => {
          update(changes?.resources)
        }}
        mock={true}
        readonly={true}
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })
  it('renders as expected when a dependency namespace is invalid', () => {
    const update = jest.fn()
    const Component = (props) => {
      return <SyncEditor {...props} />
    }
    const { asFragment } = render(
      <Component
        editorTitle={'Policy YAML'}
        variant="toolbar"
        resources={[
          {
            apiVersion: 'policy.open-cluster-management.io/v1',
            kind: 'Policy',
            metadata: {
              name: 'policy-pod',
              namespace: 'default',
            },
            spec: {
              dependencies: [
                {
                  apiVersion: 'policy.open-cluster-management.io/v1',
                  compliance: 'Compliant',
                  kind: 'ConfigurationPolicy',
                  name: 'case-test-policy',
                  namespace: 'default',
                },
              ],
              disabled: false,
              'policy-templates': [
                {
                  objectDefinition: {
                    apiVersion: 'policy.open-cluster-management.io/v1',
                    kind: 'ConfigurationPolicy',
                    metadata: {
                      name: 'policy-pod-1',
                    },
                    spec: {},
                  },
                },
              ],
            },
          },
        ]}
        schema={schema}
        onEditorChange={(changes) => {
          update(changes?.resources)
        }}
        mock={true}
        readonly={true}
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
