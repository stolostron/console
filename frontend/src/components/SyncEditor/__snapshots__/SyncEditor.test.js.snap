// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SyncEditor component renders as expected 1`] = `
<DocumentFragment>
  <div
    class="sync-editor__container"
  >
    <div
      class="pf-c-code-editor pf-m-read-only"
    >
      <div
        class="pf-c-code-editor__header"
      >
        <div
          class="pf-c-code-editor__controls"
        >
          <div
            class="sy-c-code-editor__title"
          >
            Policy YAML
          </div>
          <div
            style="display: flex;"
          >
            <button
              aria-disabled="false"
              aria-label="Find"
              class="pf-c-button pf-m-control"
              data-ouia-component-id="OUIA-Generated-Button-control-1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                />
              </svg>
            </button>
            <button
              aria-disabled="false"
              aria-label="Copy to clipboard"
              aria-labelledby="copy-button code-content"
              class="pf-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              id="copy-button"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 448 512"
                width="1em"
              >
                <path
                  d="M320 448v40c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24V120c0-13.255 10.745-24 24-24h72v296c0 30.879 25.121 56 56 56h168zm0-344V0H152c-13.255 0-24 10.745-24 24v368c0 13.255 10.745 24 24 24h272c13.255 0 24-10.745 24-24V128H344c-13.2 0-24-10.8-24-24zm120.971-31.029L375.029 7.029A24 24 0 0 0 358.059 0H352v96h96v-6.059a24 24 0 0 0-7.029-16.97z"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="pf-c-code-editor__header-main"
        />
      </div>
      <div
        class="pf-c-code-editor__main"
      >
        <div
          class="pf-c-code-editor__code"
          tabindex="0"
        >
          <textarea
            aria-label="monaco"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SyncEditor component renders as expected when a dependency compliance is invalid 1`] = `
<DocumentFragment>
  <div
    class="sync-editor__container"
  >
    <div
      class="pf-c-code-editor pf-m-read-only"
    >
      <div
        class="pf-c-code-editor__header"
      >
        <div
          class="pf-c-code-editor__controls"
        >
          <div
            class="sy-c-code-editor__title"
          >
            Policy YAML
          </div>
          <div
            style="display: flex;"
          >
            <button
              aria-disabled="false"
              aria-label="Find"
              class="pf-c-button pf-m-control"
              data-ouia-component-id="OUIA-Generated-Button-control-2"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                />
              </svg>
            </button>
            <button
              aria-disabled="false"
              aria-label="Copy to clipboard"
              aria-labelledby="copy-button code-content"
              class="pf-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-2"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              id="copy-button"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 448 512"
                width="1em"
              >
                <path
                  d="M320 448v40c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24V120c0-13.255 10.745-24 24-24h72v296c0 30.879 25.121 56 56 56h168zm0-344V0H152c-13.255 0-24 10.745-24 24v368c0 13.255 10.745 24 24 24h272c13.255 0 24-10.745 24-24V128H344c-13.2 0-24-10.8-24-24zm120.971-31.029L375.029 7.029A24 24 0 0 0 358.059 0H352v96h96v-6.059a24 24 0 0 0-7.029-16.97z"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="pf-c-code-editor__header-main"
        />
      </div>
      <div
        class="pf-c-code-editor__main"
      >
        <div
          class="pf-c-code-editor__code"
          tabindex="0"
        >
          <textarea
            aria-label="monaco"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SyncEditor component renders as expected when a dependency namespace is invalid 1`] = `
<DocumentFragment>
  <div
    class="sync-editor__container"
  >
    <div
      class="pf-c-code-editor pf-m-read-only"
    >
      <div
        class="pf-c-code-editor__header"
      >
        <div
          class="pf-c-code-editor__controls"
        >
          <div
            class="sy-c-code-editor__title"
          >
            Policy YAML
          </div>
          <div
            style="display: flex;"
          >
            <button
              aria-disabled="false"
              aria-label="Find"
              class="pf-c-button pf-m-control"
              data-ouia-component-id="OUIA-Generated-Button-control-3"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                />
              </svg>
            </button>
            <button
              aria-disabled="false"
              aria-label="Copy to clipboard"
              aria-labelledby="copy-button code-content"
              class="pf-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-3"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              id="copy-button"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 448 512"
                width="1em"
              >
                <path
                  d="M320 448v40c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24V120c0-13.255 10.745-24 24-24h72v296c0 30.879 25.121 56 56 56h168zm0-344V0H152c-13.255 0-24 10.745-24 24v368c0 13.255 10.745 24 24 24h272c13.255 0 24-10.745 24-24V128H344c-13.2 0-24-10.8-24-24zm120.971-31.029L375.029 7.029A24 24 0 0 0 358.059 0H352v96h96v-6.059a24 24 0 0 0-7.029-16.97z"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="pf-c-code-editor__header-main"
        />
      </div>
      <div
        class="pf-c-code-editor__main"
      >
        <div
          class="pf-c-code-editor__code"
          tabindex="0"
        >
          <textarea
            aria-label="monaco"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
